internal/detector/kafkabroker/extract/attributes.go

```diff
@@ -0,0 +1,223 @@
+// files.
+type brokerInfo struct {
+	serverProperties string
+	logDirs          string
+	brokerID         string
+}
+
+// ShouldParseProperties returns true if the properties file should be read to extract the log directories or broker ID.
+func (i *brokerInfo) ShouldParseProperties() bool {
+	return len(i.serverProperties) != 0 && !(i.HasLogDirs() && i.HasBrokerID())
+}
+
+func (i *brokerInfo) HasLogDirs() bool {
+	return len(i.logDirs) > 0
+}
+
+func (i *brokerInfo) HasBrokerID() bool {
+	return len(i.brokerID) > 0
+}
+
+type attributesExtractor struct {
+	logger *slog.Logger
+}
+
+func NewAttributesExtractor(logger *slog.Logger) detector.Extractor[map[string]string] {
+	return &attributesExtractor{
+		logger: logger,
+	}
+}
+
+// Extract Kafka broker attributes (broker.id and cluster.id) from a process. It parses the command line arguments to
+// identify Kafka brokers, reads configuration from the properties file if needed, and reads meta.properties files in
+// Kafka log directories for cluster information. Returns a map containing non-empty attribute values.
+func (e *attributesExtractor) Extract(ctx context.Context, process detector.Process) (map[string]string, error) {
+	args, err := process.CmdlineSliceWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := e.parseArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// only parse the server properties if needed
+	if info.ShouldParseProperties() {
+		if err = e.parseServerPropertiesFile(ctx, process, info); err != nil {
+			e.logger.Debug("Failed to parse Kafka server properties", "pid", process.PID(), "err", err)
+		}
+	}
+
+	attributes := e.extractAttributesFromMetaProperties(ctx, process, info)
+	if _, ok := attributes[propertyBrokerID]; !ok && info.HasBrokerID() {
+		attributes[propertyBrokerID] = info.brokerID
+	}
+	for key, value := range attributes {
+		if value == "" {
+			delete(attributes, key)
+		}
+	}
+	return attributes, nil
+}
+
+// extractAttributesFromMetaProperties reads the generated meta.properties files in Kafka log directories to find
+// broker.id and cluster.id values.
+func (e *attributesExtractor) extractAttributesFromMetaProperties(ctx context.Context, process detector.Process, info *brokerInfo) map[string]string {
+	attributes := make(map[string]string)
+	if !info.HasLogDirs() {
+		return attributes
+	}
+	logDirs := strings.Split(info.logDirs, logDirsSeparator)
+	for _, logDir := range logDirs {
+		logDir = strings.TrimSpace(logDir)
+		if logDir == "" {
+			continue
+		}
+		dir, err := util.AbsPath(ctx, process, logDir)
+		if err != nil {
+			continue
+		}
+		err = e.parsePropertiesFile(filepath.Join(dir, fileNameMetaProperties), func(key, value string) bool {
+			switch key {
+			case propertyBrokerID, propertyNodeID:
+				attributes[propertyBrokerID] = value
+			case propertyClusterID:
+				attributes[key] = value
+			}
+			return len(attributes) < 2
```

Comment:
Any reason we wanna do all or nothing?

Reply:
It's for an early exit. I can try to make it clearer.